### PR TITLE
[preferential] Implement EP for hyperparameter fitting

### PR DESCRIPTION
**Pull request by contramundum53**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## What does this implement/fix? Explain your changes.

This PR implements EP to approximate posterior skew GP and fit hyperparameters (e.g. kernel lengthscale and noise variance).

### CODE REVIEW

File: optuna_dashboard/preferential/samplers/gp.py [hunk 1]
```diff
@@ -1,155 +1,41 @@
...
+def _orthants_MVN_Gibbs_sampling(cov_inv: Tensor, cycles: int, initial_sample: Tensor) -> Tensor:
     dim = cov_inv.shape[0]
     assert cov_inv.shape == (dim, dim)
 
-    if initial_sample is None:
-        sample_chain = np.zeros(dim)
-    else:
-        sample_chain = initial_sample
+    sample_chain = initial_sample
+    conditional_std = 1 / torch.sqrt(torch.diag(cov_inv))
```
Comment (not522) [inline optuna_dashboard/preferential/samplers/gp.py:28]:
```suggestion
    conditional_std = torch.rsqrt(torch.diag(cov_inv))
```

File: optuna_dashboard/preferential/samplers/gp.py [hunk 2]
```diff
@@ -158,144 +44,236 @@ def _orthants_MVN_Gibbs_sampling(
...
+        return 1
+
+
+def _truncnorm_mean_var_logz(alpha: Tensor) -> tuple[Tensor, Tensor, Tensor]:
+    SQRT_HALF = math.sqrt(0.5)
+    SQRT_HALF_PI = math.sqrt(0.5 * math.pi)
+    logz = torch.special.log_ndtr(-alpha)
+    mean = 1 / (SQRT_HALF_PI * torch.special.erfcx(alpha * SQRT_HALF))
+    var = 1 - mean * (mean - alpha)
+    return (mean, var, logz)
```
Comment (not522) [inline optuna_dashboard/preferential/samplers/gp.py:153]:
```suggestion
    return mean, var, logz
```

File: optuna_dashboard/preferential/samplers/gp.py [hunk 3]
```diff
@@ -158,144 +44,236 @@ def _orthants_MVN_Gibbs_sampling(
...
+            db = (mean1 * var2 - mean2 * var1) * var1_var2_inv
+            da = (var1 - var2) * var1_var2_inv
+            virtual_obs_b[i] = virtual_obs_b[i] + db
+            virtual_obs_a[i] = virtual_obs_a[i] + da
+
+            dr = (1 + var1 * da).reciprocal()
+            mu = mu - Sxy * ((db + mean1 * da) * dr)
+            cov = cov - (Sxy[:, None] * (da * dr)) @ Sxy[None, :]
+            log_zs[i] = logz
+    return (mu, cov, torch.sum(log_zs))
```
Comment (not522) [inline optuna_dashboard/preferential/samplers/gp.py:197]:
```suggestion
    return mu, cov, torch.sum(log_zs)
```

File: optuna_dashboard/preferential/samplers/gp.py [hunk 4]
```diff
@@ -158,144 +44,236 @@ def _orthants_MVN_Gibbs_sampling(
...
+                if torch.max(torch.abs(p_old - p_new)) > tolerance:
+                    break
+            else:
+                break
+            last_params = [p.detach().clone() for p in params]
+
+    def sample_gp(self, x: Tensor, preferences: Tensor) -> _SampledGP:
+        self.fit_params_EP(x, preferences)
+        print({name: p.exp() for name, p in self.kernel.named_parameters()})
+        print({"noise": self.log_noise.exp()})
```
Comment (not522) [inline optuna_dashboard/preferential/samplers/gp.py:252]: Could you change them to logging messages?

**Review by not522 — COMMENTED** [review]
Sorry for my late response. I added minor comments.

File: optuna_dashboard/preferential/samplers/gp.py (hunk 4 above)
Comment (contramundum53) [reply optuna_dashboard/preferential/samplers/gp.py:252]: Sorry, this is not intentional. I'll remove these prints.

**Review by not522 — APPROVED** [review]
LGTM!

**Pull request merged**